lib/webdriver.py

```diff
@@ -260,6 +314,50 @@ def _get_needed_headers(self, request_headers: JSON) -> JSON:
 
         return headers
 
+    def _is_cache_expired(self, cached_time_str: str, current_time: datetime) -> bool:
+        try:
+            cached_time = datetime.fromisoformat(cached_time_str)
+            return current_time - cached_time > self.cache_duration
+        except ValueError:
+            logger.error(f"Invalid timestamp format: {cached_time_str}")
+            return True
+
+    def _cached_headers_manager(
+        self,
+        current_time: Optional[datetime] = None,
+        login_failed: Optional[bool] = None,
+        headers: Optional[JSON] = None,
+    ) -> JSON:
+        """
+        Read existing cached data from JSON file or update it with new values.
+        """
+        if os.path.exists(self.cached_headers_path):
+            try:
+                with open(self.cached_headers_path, "r") as file:
+                    cached_data = json.load(file)
+            except json.JSONDecodeError:
+                logger.error(f"Error decoding JSON from {self.cached_headers_path}.")
+                cached_data = {}
+        else:
+            cached_data = {}
+
+        # Update cached data if new values are provided
+        if current_time:
+            cached_data["timestamp"] = current_time.isoformat()
+        if login_failed is not None:
+            cached_data["login_failed"] = login_failed
+        if headers:
+            cached_data["headers"] = headers
+
+        # Write the updated data back to the JSON file
+        try:
+            with open(self.cached_headers_path, "w") as file:
+                json.dump(cached_data, file, indent=4)
+        except IOError as e:
```

Comment:
Have you seen an IOError reached here? If not, we should remove this. I've never seen this fail before with this exception. Also, nothing is controlled by the user, so if there is an error I'd prefer to know about it and fix it. People have a better chance of reporting exceptions that crash the script rather than a logged error.